recipes/mender-client4/debian-master/control

```diff
@@ -7,37 +7,36 @@ Build-Depends-Indep: cmake, debhelper (>= 10)
 Build-Depends: liblmdb-dev (>= 0.9.7), libc6-dev (>= 2.14), libboost-dev (>= 1.67.0), libboost-filesystem-dev (>= 1.67.0), libboost-log-dev (>= 1.67.0), libboost-regex-dev (>= 1.67.0), libboost-chrono-dev (>= 1.67.0), libboost-thread-dev (>= 1.67.0), libboost-atomic-dev (>= 1.67.0),  libssl-dev (>= 1.1.1), libarchive-dev (>= 3.2.2), libdbus-1-dev, libsystemd-dev
 Homepage: https://mender.io
 
-Package: mender-client
-Conflicts: mender
+Package: mender-client4
+Conflicts: mender, mender-client
+Replaces: mender-client, mender-auth, mender-update, mender-snapshot, mender-setup
```

Comment:
Does it replace mender-update, mender-auth, mender-setup and mender-snapshot? This doesn't seem right...

Reply:
The goal of this pacakge is to replace `mender-client`. However when we modified the package for the upgrade from 4.0.0 to 1:3.5.2, we introduce all these packages as "provides". See [here](https://github.com/mendersoftware/mender-dist-packages/commit/4a4af5c75bf1d5ba0c3d577a7e97d31706c0f283#diff-8ab54d0e6396495abbd8349b3f4bf813b79a93d881f3c4a2d1eca1930895e7c1R11).

I think we need to live with this, I tried many things and couldn't make it work otherwise.

Reply:
Mmmm. I think I tried this manually (an automated test for this is very hard) but maybe not in the last iteration. I will verify this again.

Reply:
> an automated test for this is very hard

Is it? Why?

I think `apt install mender-update mender-client4` is acceptable for the major upgrade, but I think it would be bad if you need to keep upgrading it like that also afterwards.

Reply:
> > an automated test for this is very hard
> 
> Is it? Why?
> 

Because you need two builds and we don't have an easy test framework for that. For example building `mender-client4` for tag `4.0.0` and for master. Then you first install the former and upgrade to the later.

> I think `apt install mender-update mender-client4` is acceptable for the major upgrade, but I think it would be bad if you need to keep upgrading it like that also afterwards.

I agree that that would be very bad. I however just verified that it works for `apt upgrade`:

```
vagrant@bullseye:~$ sudo apt upgrade
...
Preparing to unpack .../3-mender-auth_4.1.0~git20240206.f4459226-1+debian+buster+builderLOCAL_amd64.deb ...
Unpacking mender-auth (4.1.0~git20240206.f4459226-1+debian+buster+builderLOCAL) over (4.0.0-1+debian+buster) ...
Preparing to unpack .../4-mender-update_4.1.0~git20240206.f4459226-1+debian+buster+builderLOCAL_amd64.deb ...
Unpacking mender-update (4.1.0~git20240206.f4459226-1+debian+buster+builderLOCAL) over (4.0.0-1+debian+buster) ...
Preparing to unpack .../5-mender-client4_4.1.0~git20240206.f4459226-1+debian+buster+builderLOCAL_amd64.deb ...
Unpacking mender-client4 (4.1.0~git20240206.f4459226-1+debian+buster+builderLOCAL) over (4.0.0-1+debian+buster) ...
```

And for `apt install`:

```
vagrant@bullseye:~$ sudo apt install mender-client4
...
Preparing to unpack .../mender-client4_4.1.0~git20240206.f4459226-1+debian+buster+builderLOCAL_amd64.deb ...
Unpacking mender-client4 (4.1.0~git20240206.f4459226-1+debian+buster+builderLOCAL) over (4.0.0-1+debian+buster) ...
```

Reply:
> Because you need two builds and we don't have an easy test framework for that. For example building `mender-client4` for tag `4.0.0` and for master. Then you first install the former and upgrade to the later.

My thinking was that you can just reinstall the same version. So not technically an upgrade. Granted, I'm not *certain* that this would trigger the same behavior in dpkg, but it's better than nothing, and should be pretty easy to test.

> I however just verified that it works for `apt upgrade`:

👍

Reply:
> My thinking was that you can just reinstall the same version. So not technically an upgrade. Granted, I'm not _certain_ that this would trigger the same behavior in dpkg, but it's better than nothing, and should be pretty easy to test.
> 

Right, I didn't think about it. If it would have been an issue, I am not sure it would have shown with a `reinstall` command.

In any case it works (tested manually) and once this is in production it will be tested in regression with [this other test](https://github.com/mendersoftware/mender-dist-packages/blob/75a93c1cf33905488cba58a15709c54769fc55a9/tests/test_install_mender_sh.py#L228), so I think we are good here.